### PR TITLE
[ADD] translation-positional-used: New check to valid the allow change the order of string formatting

### PR DESCRIPTION
**Pull request by moylop260**
If you are using:
 - `_("Your %s %s is shipping") % ("red", "car")`

Better using:
 - `_("Your %(color)s %(item)s is shipping") % {"color": "red", "item": "car"}`

In order to allow other languages change the order
 - e.g. in Spanish: `"Su %(item)s %(color)s está enviándose"`

Fix https://github.com/OCA/pylint-odoo/issues/303

### CODE REVIEW

**Comment by Yajo** [comment]
There's something wrong here. You're also checking strings such as `"Your {} {} is shipping"`. But that's not a problem because:

1. If you're using format string syntax, you would be able to use `"Su {1} {0} está enviándose"`.
1. That syntax should be instead forbidden completely due to security problems as explained in #304.

**Comment by moylop260** [comment]
@Yajo 
Good point!

Could you check it again, please?

I just considered the `printf` cases
The format ones could be changed as you said but avoid from the another lint-check (no this one).

Thanks!

File: pylint_odoo/checkers/no_modules.py [hunk 1]
```diff
@@ -605,11 +612,23 @@ def visit_call(self, node):
...
                 self.add_message(
                     'translation-contains-variable', node=node,
                     args=(wrong, right))
 
+            # translation-positional-used: Check "string to translate"
+            # to check "%s %s..." used where the position can't be changed
+            str2translate = arg.as_string()
+            printf_args = (
+                misc.WrapperModuleChecker.
+                _get_printf_str_args_kwargs(str2translate))
```
Comment (Yajo) [inline pylint_odoo/checkers/no_modules.py:625]:
```suggestion
                misc._get_printf_str_args_kwargs(str2translate))
```

File: pylint_odoo/checkers/no_modules.py [hunk 2]
```diff
@@ -226,6 +226,12 @@
         'print-used',
         settings.DESC_DFLT
     ),
+    'W%d20' % settings.BASE_NOMODULE_ID: (
+        'Translation method _(%s) is using positional string printf formatting. '
+        'Use named placeholder "_(%%(placeholder)s)" instead.',
```
Comment (Yajo) [inline pylint_odoo/checkers/no_modules.py:231]:
```suggestion
        'Use named placeholder `_("%%(placeholder)s")` instead.',
```

I guess the double-percent symbol is needed, right?

File: pylint_odoo/test_repo/broken_module/models/broken_model.py [hunk 3]
```diff
@@ -245,6 +245,23 @@ def my_method1(self, variable1):
         error_msg = _('Variable not translatable: {}'.format(variable1))
         error_msg = _('Variables not translatable: {}, {variable2}'.format(
             variable1, variable2=variable2))
+
+        # string with parameters without name
+        # so you can't change the order in the translation
+        _('%s %d') % ('hello', 3)
+        _('%s %s') % ('hello', 'world')
+        _('%s %s %s') % ('hello', 'world!')
```
Comment (Yajo) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py:253]: This seems like a different failure. I'm not sure if normal pylint would detect it. You're using 3 paceholders with 2 args.

Possibly not interesting to implement, given in this case you should use named placeholders.

File: pylint_odoo/test_repo/broken_module/models/broken_model.py [hunk 4]
```diff
@@ -245,6 +245,23 @@ def my_method1(self, variable1):
...
+        # Valid cases
+        _('{0} {1}') % ('hello', 3)
+        _('%(strname)s') % {'strname': 'hello'}
+        _('%(strname)s %(intname)d') % {'strname': 'hello', 'intname': 3}
+        _('%s') % 'hello'
+        _('%d') % 3
+        _('{}') % 'hello'
+        _('{:d}') % 3
+        _('{} {}') % ('hello', 3)  # It can be translated to '{1} {0}'
+        _('{} {:d}') % ('hello', 3)
```
Comment (Yajo) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py:264]: Are these tests worthwhile?

File: pylint_odoo/test_repo/broken_module/models/broken_model.py [hunk 5]
```diff
@@ -245,6 +245,23 @@ def my_method1(self, variable1):
...
             variable1, variable2=variable2))
+
+        # string with parameters without name
+        # so you can't change the order in the translation
+        _('%s %d') % ('hello', 3)
+        _('%s %s') % ('hello', 'world')
+        _('%s %s %s') % ('hello', 'world!')
+
+        # Valid cases
+        _('{0} {1}') % ('hello', 3)
```
Comment (Yajo) [inline pylint_odoo/test_repo/broken_module/models/broken_model.py:256]: Is it worth it?

File: pylint_odoo/checkers/no_modules.py (hunk 2 above)
Comment (fernandahf) [reply pylint_odoo/checkers/no_modules.py]: Done

File: pylint_odoo/checkers/no_modules.py (hunk 1 above)
Comment (fernandahf) [reply pylint_odoo/checkers/no_modules.py:631]: Hi @Yajo 

This is full line:

```
misc.WrapperModuleChecker._get_printf_str_args_kwargs(str2translate))
```
`_get_printf_str_args_kwargs` comes from `WrapperModuleChecker` class.

File: pylint_odoo/test_repo/broken_module/models/broken_model.py (hunk 3 above)
Comment (fernandahf) [reply pylint_odoo/test_repo/broken_module/models/broken_model.py]: I think that is a different failure. I removed it.

File: pylint_odoo/test_repo/broken_module/models/broken_model.py (hunk 4 above)
Comment (fernandahf) [reply pylint_odoo/test_repo/broken_module/models/broken_model.py]: I think a few examples is enough.

File: pylint_odoo/test_repo/broken_module/models/broken_model.py (hunk 5 above)
Comment (fernandahf) [reply pylint_odoo/test_repo/broken_module/models/broken_model.py:255]: I think a few examples is enough. I removed a few examples.

File: pylint_odoo/checkers/no_modules.py (hunk 1 above)
Comment (Yajo) [reply pylint_odoo/checkers/no_modules.py:631]: Oh sorry, the formatting tricked me. 🤦🏼‍♂️

File: pylint_odoo/test_repo/broken_module/models/broken_model.py (hunk 5 above)
Comment (Yajo) [reply pylint_odoo/test_repo/broken_module/models/broken_model.py:255]: This line is not a valid case. It uses str-format syntax with `%`

Comment (fernandahf) [reply pylint_odoo/test_repo/broken_module/models/broken_model.py]: You right, I removed it.

**Review by Yajo — APPROVED** [review]

**Comment by fernandahf** [comment]
@luisg123v 

Could you review this, please?

Regards.

**Pull request merged**